File: java/src/main/java/problems301_400/IntersectionOfTwoArrays2.java
package problems301_400;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

enum IntersectionOfTwoArrays2 {
    DEFAULT {
        @Override
        int[] intersect(int[] nums1, int[] nums2) {
            Arrays.sort(nums1);
            Arrays.sort(nums2);

            int i = 0;
            int j = 0;
            List<Integer> result = new ArrayList<>();
            while (i < nums1.length && j < nums2.length) {
                if (nums1[i] < nums2[j]) {
                    i++;
                } else if (nums1[i] > nums2[j]) {
                    j++;
                } else {
                    result.add(nums1[i]);
                    i++;
                    j++;
                }
            }

            int[] intersect = new int[result.size()];
            for (int k = 0; k < intersect.length; k++) {
                intersect[k] = result.get(k);
            }
            return intersect;
        }
    };

    abstract int[] intersect(int[] nums1, int[] nums2);
}

File: kotlin/src/main/kotlin/problems301_400/IntersectionOfTwoArrays2.kt
package problems301_400

import java.util.Arrays

enum class IntersectionOfTwoArrays2 {
    DEFAULT {
        override fun intersect(nums1: IntArray, nums2: IntArray): IntArray {
            Arrays.sort(nums1)
            Arrays.sort(nums2)

            var i = 0
            var j = 0
            val result: MutableList<Int> = ArrayList()
            while (i < nums1.size && j < nums2.size) {
                if (nums1[i] < nums2[j]) {
                    i++
                } else if (nums1[i] > nums2[j]) {
                    j++
                } else {
                    result.add(nums1[i])
                    i++
                    j++
                }
            }

            val intersect = IntArray(result.size)
            for (k in intersect.indices) {
                intersect[k] = result[k]
            }
            return intersect
        }
    },
    ;

    abstract fun intersect(nums1: IntArray, nums2: IntArray): IntArray
}
